### ISSUE TITLE
Design a "Sensor Fusion" Layer To Separate Network Input And Sensor Fusion

### ISSUE THREAD

**Issue by garethellis0** [UBC-Thunderbots/Software#764]
### Description of the task
- We should logically separate reading raw vision data from filtering
- Filtering should also read radio breakbeam data (more generally any data from the robots) to help know where the ball is
- Removing expired ball and team data should be done regardless of cameras being enabled / zones being disabled, so we remove things properly and keep updating if we disable a camera. 
- General idea here is we need to have a separate layer to integrate data from both the robots and network.

**This is only creating the new layer, not adding any new algorithms**. Things like the ball and robot filters should just need to be moved and shouldn't need to be changed.

The data being passed from the `Backend` to the `SensorFusion` to be our simple structs of `BallDetection`, `RobotDetection`, `RefboxData` etc.

### Acceptance criteria

- [ ] Create an Observer/Subject for "SensorFusion" See the image below for the interface.
- [ ] Move all filtering from the backend to the sensor fusion

### Blocked By

#1065 is needed to handle refbox data

**Comment by MathewMacDougall** [comment]
![IMG_20191116_134428671_HDR](https://user-images.githubusercontent.com/21165407/68999649-29e20f00-0878-11ea-87de-4649d66e02fc.jpg)

**Comment by jonathanlew** [comment]
:falkland_islands: :

**Comment by MathewMacDougall** [comment]
After discussion with @jonathanlew this issue is blocked by #1065 . We will merge whatever has currently been done  that's useful (folder structure, etc) but otherwise this issue will be on hold for now as we focus on higher-priorities like mrftest (and work on unblocking this issue).
